cluster-api/machines-api/types.go

```diff
@@ -0,0 +1,160 @@
+	LastUpdated metav1.Time
+
+	// Indicates whether or not the Machine is fully reconciled. When a
+	// controller observes that the spec has changed and no longer matches
+	// reality, it should update Ready to false before reconciling the
+	// state, and then set back to true when the state matches the spec.
+	Ready bool
+
+	// If set, indicates that there is a problem reconciling state, and
+	// will be set to a token value suitable for machine interpretation.
+	// +optional
+	ErrorReason *MachineStatusError
+
+	// +optional
+	// If set, indicates that there is a problem reconciling state, and
+	// will be set to a human readable string to indicate the problem.
+	ErrorMessage *string
+}
+
+type MachineStatusError string
+
+const (
+	// Represents that the combination of configuration in the MachineSpec
+	// is not supported by this cluster. This is not a transient error, but
+	// indicates a state that must be fixed before progress can be made.
+	//
+	// Example: the ProviderConfig specifies an instance type that doesn't exist,
+	InvalidConfigurationMachineError MachineStatusError = "InvalidConfiguration"
+
+	// This indicates that the MachineSpec has been updated in a way that
+	// is not supported for reconciliation on this cluster. The spec may be
+	// completely valid from a configuration standpoint, but the controller
+	// does not support changing the real world state to match the new
+	// spec.
+	//
+	// Example: the responsible controller is not capable of changing the
+	// container runtime from docker to rkt.
+	UnsupportedChangeMachineError MachineStatusError = "UnsupportedChange"
+
+	// This generally refers to exceeding one's quota in a cloud provider,
+	// or running out of physical machines in an on-premise environment.
+	InsufficientResourcesMachineError MachineStatusError = "InsufficientResources"
+
+	// There was an error while trying to create a Node to match this
+	// Machine. This may indicate a transient problem that will be fixed
+	// automatically with time, such as a service outage, or a terminal
+	// error during creation that doesn't match a more specific
+	// MachineStatusError value.
+	//
+	// Example: timeout trying to connect to GCE.
+	CreateMachineError MachineStatusError = "CreateError"
+
+	// An error was encountered while trying to delete the Node that this
+	// Machine represents. This could be a transient or terminal error, but
+	// will only be observable if the provider's Machine controller has
+	// added a finalizer to the object to more gracefully handle deletions.
+	//
+	// Example: cannot resolve EC2 IP address.
+	DeleteMachineError MachineStatusError = "DeleteError"
+)
+
+type MachineVersionInfo struct {
+	// Semantic version of kubelet to run
+	Kubelet string
+
+	// Name/version of container runtime
+	ContainerRuntime ContainerRuntimeInfo
+}
+
+type ContainerRuntimeInfo struct {
+	// docker, rkt, containerd, ...
+	Name string
+
+	// Semantic version of the container runtime to use
+	Version string
```

Comment:
Going to guess these should be optional to mean "use the controller-recommended setting for the k8s/kubelet version" (which IMO is what we should be encouraging!)

Reply:
Ah, good point. I think it's a good idea to make these optional at cluster creation time, but do you think they should stay optional afterwards?

Reply:
Be careful with default values for fields. Changing defaults can easily break an API and unlike ProviderConfig there's no clear provider-specific versioning mechanism for the defaults applied to these common fields.